node/src/main/kotlin/net/corda/node/services/network/NetworkMapUpdater.kt

```
@@ -136,6 +137,9 @@ class NetworkMapUpdater(private val networkMapCache: NetworkMapCacheInternal,
     private fun watchForNodeInfoFiles(): Subscription {
         return nodeInfoWatcher
                 .nodeInfoUpdates()
+                .doOnError { logger.warn("Error encountered while polling directory for network map updates, " +
+                        "will retry in $defaultWatchNodeInfoFilesRetryIntervalSeconds seconds - $it") }
+                .retryWhen { t -> t.delay(defaultWatchNodeInfoFilesRetryIntervalSeconds, TimeUnit.SECONDS, nodeInfoWatcher.scheduler) }
```

Some thoughts to better understand the proposed solution:

1. I think the body of the `retryWhen` can be shortened:

```
.retryWhen { t -> t.delay(defaultWatchNodeInfoFilesRetryIntervalSeconds, TimeUnit.SECONDS) }  

```

and the visibility of the scheduler can be reverted to private. Why is access to the scheduler in the nodeInfoWatcher necessary?

2. Most of the documented approaches use a flatMap inside the `retryWhen`:

```
.retryWhen { errors ->
                    logger.warn("Error encountered while polling directory for network map updates, " +
                            "will retry in $defaultWatchNodeInfoFilesRetryIntervalSeconds seconds")
                    errors.flatMap { Observable.timer(defaultWatchNodeInfoFilesRetryIntervalSeconds, TimeUnit.SECONDS) }
                }
```

 - This allows for the logging to also be performed inside the `retryWhen` body.
 - Allows to defer the decision until you know what kind of error you are dealing with. This means that if NoSuchFileException is thrown, will be handled differently in case if there is a different exception.


3. I think if the `retryWhen{}` is used, then the observer will be re-subscribed to a new observable until the expected event happens.  A counter may be used to keep track of number of retries and stop retrying after a certain number of times.

---

The `scheduler` is for testing purposes,to advance time and assert the observable picks up file changes after the delay. I'm not sure of a better way, I'm not sure if we can advance time on the default scheduler (`computationScheduler` for `delay()`)

Logging in `doOnError` is just a way to separate concerns, but I do like your idea, currently, certain exceptions could be swallowed and we'll never know what happened.

It is quite a lot of warnings in the output logs, I think some incrementing backoff would be a good idea, maybe adding 10s every time to a max of 1 minute or so. Capping the retries will mean `onError` gets invoked, Observable will get removed and node will have to be restarted to pick up additional-node-info updates. 

So how about if I do the following?
1. in retryWhen, add more generic error message and the stacktrace for other exceptions
2. add an incrementing backoff (+10 seconds capped at 60 seconds delay) to reduce spam on the logs

---

I assume the retry with increasing backoff fun was never added? Then why do we need the previousConsecutiveErrors at all?

---

No backoff logic was not added. The previousConsecutiveErrors could be named better, it just keeps track of there being a previous error so we only log the first error, and only log the first succeeded, and if further errors happen we log them too.